### ISSUE TITLE
Member function definition is not listed in outline

### ISSUE THREAD

**Issue by Yanpas** [MaskRay/ccls#169]
Here is my file. If I place cursor inside S::method definiton then breadcrumbs won't tell me if I'm inside method definition.

![image](https://user-images.githubusercontent.com/7345761/50284954-ba974b80-046b-11e9-8133-4066eba69b86.png)

Src:
```cpp
namespace r{
  void foo() {

  }

  struct S {
    int a;
    S() {
      a = 5;
    }
    void method() const;
  };
  void S::method() const {
    puts("");
  }

}
```

**Comment by MaskRay** [comment]
> then breadcrumbs won't tell me if I'm inside method definition.

Please clarify what the UI feature is as not everyone uses VSCode. I presume `range` makes this work:

```
	/**
	 * The range enclosing this symbol not including leading/trailing whitespace but everything else
	 * like comments. This information is typically used to determine if the clients cursor is
	 * inside the symbol to reveal in the symbol in the UI.
	 */
	range: Range;
```

```javascript
{"name":"method","detail":"void r::S::method() const","kind":6,
"range":{"start":{"line":10,"character":4},"end":{"line":10,"character":23}},
"selectionRange":{"start":{"line":10,"character":9},"end":{"line":10,"character":15}},
"children":[]}]}
```

`range` describe the extent of `void method() const;` (line 10, 0-based).

`r::S::method` has both a declaration and a definition. This is a tricky case as `range` cannot describe the two extents while both may be useful.

**Comment by Yanpas** [comment]
Yes, I think it's `textDocument/documentSymbol` and `DocumentSymbol.range` field. Or maybe namespace lacks child in `DocumentSymbol.children`.

> Please clarify what the UI feature is

It's called breadcrumbs, you can enable it via "breadcrumbs.enabled" setting. In terms of LSP that means most probably that either range is incorrect, either children do not include it.

There is definetely some trouble with namespaces, here is another example.
File a.h:
```cpp
namespace  xx {

class klass1 {
  void method() const;
};

} // xx

class klass2 {
  void method() const;
};

```

File a.cpp:
```cpp
#include "a.h"

#include <cstdio>

namespace  xx {

void klass1::method() const {
  puts("1");
}

} // xx

void klass2::method() const {
  puts("2");
}
```

If I place cursor on line `puts("1")` breadcrumbs show me that I'm inside namespace xx (but should that I'm inside ns xx inside method). But if I place cursor on line `puts("2")` breadcrumbs show me that I'm inside method

**Comment by MaskRay** [comment]
I am not sure whether the current specification of `interface DocumentSymbol` (the return type used by hierarchical document symbol) can make the scenario work. These redeclarable things are called `clang::Redeclarable`.

* member functions. For a function that does not have its definition in the same file, making `range` point to its declaration is a reasonable choice. For a function with both in-class declaration and out-of-line definition, for consistency `range` should point to its declaration but for the `breadcrumbs.enabled` feature it is better to use its definition. A discrepancy here.
* namespaces. `namespace ns {}` can be used multiple times in a file.
* templates.
* ObjcC interfaces and protocols.

**Comment by MaskRay** [comment]
Fixed by #309

**Comment by Yanpas** [comment]
Thx!
